### DISCUSSION TITLE
Question about some operations

### DISCUSSION THREAD

**Discussion by shy218** [sgkit-dev/sgkit#526] [Q&A]
What is the math formula for PC-Relate in relatedness estimation?

Is the LD prune supported?

**Comment by hammer** [comment]
@ravwojdyla knows the PC-Relate code the best.

LD-prune is being actively worked on right now by @tomwhite. He’s using the original implementation by @eric-czech as a template.

**Reply by shy218** [reply]
For HWE, are you using the equation from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1199378/?

**Reply by hammer** [reply]
Have you tried reading the source? https://github.com/pystatgen/sgkit/blob/master/sgkit/stats/hwe.py

**Reply by shy218** [reply]
Yes. I fully understand your implementation in HWE.

**Reply by hammer** [reply]
Great! Did you see that paper is referenced in the source then?

**Reply by shy218** [reply]
Yes, I see the reference on the paper, and I google that paper to get that link.  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1199378/. I just want to make sure whether I look into the right paper.

**Reply by hammer** [reply]
Thanks for the explanation. Yes that’s the paper. Perhaps you can submit a PR that augments the comment with a link to the paper or at least the DOI?

**Reply by shy218** [reply]
Sorry for any confusion. I also have a question for variant normalization. Does variant normalization mean to make sure the variant is parsimonious and left-aligned? Is there any implementation in sgkit? Sorry if I ask a stupid question.

**Reply by hammer** [reply]
Yes, that's roughly what variant normalization means. It might also mean "lifting" the variant onto another version of the reference genome, or a few other operations. We do not have any of those operations in `sgkit` currently.

**Reply by eric-czech** [reply]
> For HWE, are you using the equation from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1199378/?

Hey @shy218, a few more details on that:

The calculation is integrating over an exact distribution that is described pretty succinctly at  https://hail.is/docs/0.2/_static/LeveneHaldane.pdf.  I ported the implementation in the paper you mentioned from the C version at http://csg.sph.umich.edu/abecasis/Exact/index.html.  I doubt the specifics will interest you much but I am certainly curious how this one performs since I haven't compared this to Hail/PLINK (I have a vague idea how several other operations perform).  

I missed this on the call but **HWE performance depends on the data**, much like LD-prune -- I forgot to mention that.  The others we discussed are only dependent on array shapes.  It is easy to generate data for realistic or worst-case scenarios though.

**Reply by shy218** [reply]
![WeChat Photo Editor_20210419013057](https://user-images.githubusercontent.com/39325963/115154896-caf97880-a0af-11eb-98e3-7ab85c74d4c3.jpg)
I write the algorithm by steps in Math for HWE. Do you think this understanding is correct? And also, do you think it is a good idea to test on HWE compared with Hail/PLINK?

**Reply by eric-czech** [reply]
> Do you think this understanding is correct?

Not sure, you'd have to check that against the paper or the code.

> And also, do you think it is a good idea to test on HWE compared with Hail/PLINK?

By test do you mean for accuracy or for speed?  We already do the former and yes the latter would be useful.

**Comment by ravwojdyla** [comment]
👋 @shy218 see the equation 3 and 4 in https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4716688/

**Reply by shy218** [reply]
Thanks for the reply. I understand the formula now.
